File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(!head || !head->next) return nullptr ; 

        ListNode *slow = head ; 
        ListNode *fast = head ; 

        while(fast != nullptr && fast->next != nullptr && fast->next->next != nullptr){
            fast = fast->next->next ; 
            slow = slow->next ; 

            if(fast == slow){
                ListNode *temp = head ; 
                while(temp != fast){
                    temp = temp->next ; 
                    fast = fast->next ; 
                }
                return temp ; 
            }
        }

        return nullptr ; 

    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null || head.next == null) return null ;
        ListNode dummy = new  ListNode(-1001) ; 
        dummy.next = head  ;  
        ListNode fast = dummy ; 
        ListNode slow = dummy ; 

        while(fast != null && fast.next != null && fast.next.next != null){
            slow = slow.next; 
            fast = fast.next.next ; 
            if(slow == null || fast == null) return null ; 
            if(slow == fast){
                ListNode temp = head ; 
                if(temp == fast ) return temp ; 
                while(temp != fast.next ){
                    temp = temp.next ; 
                    fast = fast.next ; 
                    if(fast.next == temp) return temp ; 
                }
                return temp ; 
            }
        }
        return null ; 
    }
}